### ISSUE TITLE
Provide access to Gradle.StartParameter property value in a configuration cache compliant way

### ISSUE THREAD

**Issue by ZakTaccardi** [gradle/gradle#19474]
The following provider does not support the configuration cache.

```kts
private fun ProviderFactory.startParameterGradleProperty(
    project: Project,
    propertyName: String
): Provider<String> = provider<String?> {
        project.gradle
            .startParameter
            .projectProperties
            .get(propertyName) as String?
}
```

Unfortunately, `providers.gradleProperty(..)` does not work for me here as I need to differentiate between a gradle property passed as a start parameter and one passed in a `gradle.properties` file.

Use case is https://gist.github.com/ZakTaccardi/e2850127b94553c5f933254961c811ea - which is an attempted workaround to:
  * https://github.com/gradle/gradle/issues/12283
  * https://github.com/gradle/gradle/issues/13302

**Comment by Vampire** [comment]
Besides that there are a few unnecessary explicit types or cases, what is not configuration cache friendly with your code?
Works perfectly fine here.

**Comment by ALikhachev** [comment]
@Vampire There's a problem that with configuration cache the value of properties from start parameters is cached instead of providing the actual value. For example, we have the use-case when we access `project.gradle.startParameter.isOffline`. There's no way to mark it as configuration cache input neither to request the actual value on task execution.

**Comment by Vampire** [comment]
What has that to do with the topic here?
This is about Gradle properties from start parameters. And that works perfectly fine, as any changes in those invalidate the configuration cache.

**Comment by ALikhachev** [comment]
Yep, you're right. Previously (before Gradle 7.4) the topic was having pretty the same problem.

**Comment by ov7a** [comment]
Sorry for the late reply.

Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not in focus. It might take a while before it gets implemented.
